File: dailyQuestion/2020-07/07-10/golang/solution_dp.go
package main

import "fmt"

func maxProfit(prices []int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}
	dp := make([][3]int, n)
	dp[0][0] = -prices[0]
	for i := 1; i < n; i++ {
		dp[i][0] = max(dp[i-1][0], dp[i-1][2]-prices[i])
		dp[i][1] = dp[i-1][0] + prices[i]
		dp[i][2] = max(dp[i-1][1], dp[i-1][2])
	}
	return max(dp[n-1][1], dp[n-1][2])

}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	prices := []int{1, 2, 3, 0, 2}
	result := maxProfit(prices)
	fmt.Println(result)

}

File: dailyQuestion/2020-07/07-10/python/solution_dp.py
def maxProfit(prices: [int]) -> int:
    n = len(prices)
    if n == 0 :
        return 0
    dp0 = -prices[0]
    dp1 = 0
    dp2 = 0
    for i in range(1,n) :
        _dp0 = max(dp0,dp2-prices[i])
        _dp1 = dp0 + prices[i]
        _dp2 = max(dp1,dp2)
        dp0,dp1,dp2 = _dp0,_dp1,_dp2
    return max(dp1,dp2)

if __name__ == "__main__" :
    prices = [1,2,3,0,2]
    result = maxProfit(prices)
    print(result)
